### ISSUE TITLE
Inline host config at buildtime

### ISSUE THREAD

**Issue by jquense** [jquense/react-dom-lite#27]
we should do this: https://github.com/facebook/react/pull/12792

**Comment by gaearon** [comment]
Yeah.. I've been wondering how to get it working with custom renderers. It might be that after my changes, just running GCC would be sufficient because `$$$config` is almost not passed around. So maybe it will just inline them. If not, we could consider something like taking host instance methods as `...args` so that there's no wrapper. Then it should definitely be inlineable although it would make it harder to follow changes.

**Comment by prometheansacrifice** [comment]
@gaearon How did you manage to inline functions in [compilation_level: simple](https://github.com/facebook/react/blob/fa7fa812c70084e139d13437fb204fcdf9152299/scripts/rollup/build.js#L65). [Docs](https://developers.google.com/closure/compiler/docs/compilation_levels) mention inlining under advanced level optimisations .

Also, on a relevant note, such optimisations would need build level tests too I guess to be safe? It seems GCC can break builds if the assumptions it makes are not respected.

**Comment by gaearon** [comment]
The docs are probably outdated. GCC definitely inlines small functions even in simple mode if it determines it's safe to do. Advanced mode is more aggressive.

>such optimisations would need build level tests too I guess to be safe

Ideally yes, and we do that in React.

>It seems GCC can break builds if the assumptions it makes are not respected.

I thought so, but in practice we haven't seen a single GCC bug since adopting it. It powers Google infra so it's probably very stable.
